File: matrix_algorithms/src/matrices/ResultadoGuardador.java
package matrices;

import java.io.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

/**
 * Guarda los resultados de cada prueba en dos archivos:
 *   - resultados.csv : para graficar en Excel
 *   - resultados.txt : reporte legible con explicacion del calculo teorico
 */
public class ResultadoGuardador {

    static final String CSV = "matrix_algorithms/resultados.csv";
    static final String TXT = "matrix_algorithms/resultados.txt";
    static final DateTimeFormatter FMT = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

    public static void guardar(
            int tamano, int digitos,
            String[] nombres, long[] estimados, long[] reales,
            double k, double[] exponentes, double[] factores) throws IOException {

        String timestamp = LocalDateTime.now().format(FMT);
        boolean csvExiste = new File(CSV).exists();

        // --- CSV para Excel ---
        try (PrintWriter pw = new PrintWriter(new FileWriter(CSV, true))) {
            if (!csvExiste) {
                pw.println("Fecha,Tamano,Digitos,Algoritmo,Complejidad,Operaciones,Estimado_ms,Real_ms,Ratio");
            }
            for (int i = 0; i < nombres.length; i++) {
                long ops = Math.round(factores[i] * Math.pow(tamano, exponentes[i]));
                double ratio = (estimados[i] > 0 && reales[i] >= 0)
                    ? (double) reales[i] / estimados[i] : 0;
                pw.printf("%s,%d,%d,%s,O(n^%.3f),%d,%d,%d,%.2f%n",
                    timestamp, tamano, digitos, nombres[i],
                    exponentes[i], ops,
                    estimados[i],
                    reales[i] >= 0 ? reales[i] : -1,
                    ratio);
            }
        }

        // --- TXT reporte legible ---
        try (PrintWriter pw = new PrintWriter(new FileWriter(TXT, true))) {
            pw.println();
            pw.println(linea('=', 70));
            pw.println("PRUEBA: " + timestamp);
            pw.println("Matriz : " + tamano + " x " + tamano);
            pw.println("Digitos: " + digitos + " por valor");
            pw.println(linea('=', 70));
            pw.println();

            pw.println("COMO SE CALCULA EL TIEMPO TEORICO:");
            pw.println(linea('-', 70));
            pw.println("1. Calibracion: se mide NaivOnArray con matriz 16x16");
            pw.println("   k = tiempo_real / 16^3 = tiempo / 4096 operaciones");
            pw.printf( "   k = %.6f ms por operacion BigInteger en esta maquina%n", k);
            pw.println();
            pw.println("2. Para cada algoritmo:");
            pw.println("   operaciones = factor * n^exponente");
            pw.println("   tiempo_estimado = k * operaciones");
            pw.println();
            pw.println("   Exponentes por complejidad teorica:");
            pw.println("   NaivOnArray, bloques : O(n^3)     exponente=3.0,   factor=1.0");
            pw.println("   LoopUnrollingTwo     : O(n^3)     exponente=3.0,   factor=0.5");
            pw.println("   LoopUnrollingFour    : O(n^3)     exponente=3.0,   factor=0.25");
            pw.println("   WinogradOriginal     : O(n^3)     exponente=3.0,   factor=0.5");
            pw.println("   StrassenNaiv         : O(n^2.807) exponente=2.807, factor=1.0");
            pw.println("   StrassenWinograd     : O(n^2.807) exponente=2.807, factor=0.83");
            pw.println();
            pw.println("3. Ejemplo con n=" + tamano + ":");
            long opsEj = (long) Math.pow(tamano, 3);
            pw.printf( "   NaivOnArray: %.6f * 1.0 * %d^3 = %.6f * %d = %.0f ms%n",
                k, tamano, k, opsEj, k * opsEj);
            pw.println();

            pw.println(linea('-', 70));
            pw.printf("%-28s %-12s %-10s %-12s %-12s %-14s%n",
                "Algoritmo", "Complejidad", "Ops", "Estimado", "Real", "Ratio");
            pw.println(linea('-', 70));

            for (int i = 0; i < nombres.length; i++) {
                long ops = Math.round(factores[i] * Math.pow(tamano, exponentes[i]));
                String realStr = reales[i] >= 0 ? formatMs(reales[i]) : "ERROR";
                double ratio = (estimados[i] > 0 && reales[i] >= 0)
                    ? (double) reales[i] / estimados[i] : 0;
                String analisis = analizar(ratio);
                pw.printf("%-28s %-12s %-10s %-12s %-12s %-14s%n",
                    nombres[i],
                    "O(n^" + String.format("%.3f", exponentes[i]) + ")",
                    formatOps(ops),
                    formatMs(estimados[i]),
                    realStr,
                    reales[i] >= 0 ? String.format("%.2fx %s", ratio, analisis) : "N/A");
            }
            pw.println(linea('-', 70));
            pw.println();
            pw.println("INTERPRETACION DEL RATIO (real / estimado):");
            pw.println("  < 0.9x : mas rapido que la teoria (JIT, cache, paralelismo)");
            pw.println("  ~ 1.0x : acorde a la teoria");
            pw.println("  > 1.1x : mas lento (overhead BigInteger, GC, contention hilos)");
            pw.println();
        }

        System.out.println();
        System.out.println("Resultados guardados en:");
        System.out.println("  " + new File(CSV).getAbsolutePath());
        System.out.println("  " + new File(TXT).getAbsolutePath());
    }

    static String linea(char c, int n) {
        StringBuilder sb = new StringBuilder(n);
        for (int i = 0; i < n; i++) sb.append(c);
        return sb.toString();
    }

    static String analizar(double ratio) {
        if (ratio <= 0)   return "";
        if (ratio < 0.5)  return "(mucho mejor)";
        if (ratio < 0.9)  return "(mejor)";
        if (ratio <= 1.1) return "(acorde)";
        if (ratio <= 2.0) return "(algo lento)";
        return "(mucho mas lento)";
    }

    static String formatMs(long ms) {
        if (ms < 0)       return "ERROR";
        if (ms < 1000)    return ms + " ms";
        if (ms < 60000)   return String.format("%.2f s", ms / 1000.0);
        if (ms < 3600000) return String.format("%.2f min", ms / 60000.0);
        return String.format("%.2f h", ms / 3600000.0);
    }

    static String formatOps(long ops) {
        if (ops < 1_000)         return ops + "";
        if (ops < 1_000_000)     return String.format("%.1fK", ops / 1000.0);
        if (ops < 1_000_000_000) return String.format("%.1fM", ops / 1_000_000.0);
        return String.format("%.1fB", ops / 1_000_000_000.0);
    }
}

File: matrix_algorithms_python/resultado_guardador.py
import os
import csv
from datetime import datetime

# rutas relativas al directorio donde se ejecuta el script
CSV = "resultados.csv"
TXT = "resultados.txt"

# Cuenta cuantos casos hay ya en el CSV para numerar el nuevo
def contar_casos():
    if not os.path.exists(CSV):
        return 0
    max_caso = 0
    with open(CSV, "r") as f:
        for linea in f:
            if linea.startswith("Caso "):
                try:
                    num = int(linea.split(" ")[1])
                    if num > max_caso:
                        max_caso = num
                except:
                    pass
    return max_caso

def format_ms(ms):
    if ms < 0:       return "ERROR"
    if ms < 1000:    return f"{ms} ms"
    if ms < 60000:   return f"{ms/1000:.2f} s"
    if ms < 3600000: return f"{ms/60000:.2f} min"
    return f"{ms/3600000:.2f} h"

def format_ops(ops):
    if ops < 1_000:         return str(ops)
    if ops < 1_000_000:     return f"{ops/1000:.1f}K"
    if ops < 1_000_000_000: return f"{ops/1_000_000:.1f}M"
    return f"{ops/1_000_000_000:.1f}B"

def analizar(ratio):
    if ratio <= 0:   return ""
    if ratio < 0.5:  return "(mucho mejor)"
    if ratio < 0.9:  return "(mejor)"
    if ratio <= 1.1: return "(acorde)"
    if ratio <= 2.0: return "(algo lento)"
    return "(mucho mas lento)"

def guardar(tamano, digitos, nombres, estimados, reales, k, exponentes, factores):
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    numero_caso = contar_casos() + 1
    caso_label  = f"Caso {numero_caso} (n={tamano})"

    # crear CSV con encabezado si no existe
    escribir_header = not os.path.exists(CSV) or os.path.getsize(CSV) == 0
    with open(CSV, "a", newline="") as f:
        writer = csv.writer(f)
        if escribir_header:
            writer.writerow(["Caso","Fecha","Tamano","Digitos","Algoritmo",
                             "Complejidad","Operaciones","Estimado_ms","Real_ms","Ratio"])
        for i in range(len(nombres)):
            ops   = round(factores[i] * (tamano ** exponentes[i]))
            ratio = reales[i] / estimados[i] if estimados[i] > 0 and reales[i] >= 0 else 0
            writer.writerow([caso_label, timestamp, tamano, digitos,
                             nombres[i], f"O(n^{exponentes[i]:.3f})", ops,
                             estimados[i], reales[i] if reales[i] >= 0 else -1,
                             f"{ratio:.2f}"])

    # reporte TXT legible
    sep = "=" * 70
    lin = "-" * 70
    with open(TXT, "a") as f:
        f.write(f"\n{sep}\n")
        f.write(f"{caso_label} | {timestamp}\n")
        f.write(f"Matriz : {tamano} x {tamano}\n")
        f.write(f"Digitos: {digitos} por valor\n")
        f.write(f"{sep}\n\n")
        f.write(f"{'Algoritmo':<28} {'Complejidad':<12} {'Ops':<10} {'Estimado':<12} {'Real':<12} {'Ratio'}\n")
        f.write(f"{lin}\n")
        for i in range(len(nombres)):
            ops     = round(factores[i] * (tamano ** exponentes[i]))
            real_str = format_ms(reales[i])
            ratio   = reales[i] / estimados[i] if estimados[i] > 0 and reales[i] >= 0 else 0
            f.write(f"{nombres[i]:<28} O(n^{exponentes[i]:.3f})  {format_ops(ops):<10} "
                    f"{format_ms(estimados[i]):<12} {real_str:<12} {ratio:.2f}x {analizar(ratio)}\n")
        f.write(f"{lin}\n\n")

    print(f"\nResultados guardados como {caso_label}")
    print(f"  CSV: {os.path.abspath(CSV)}")
    print(f"  TXT: {os.path.abspath(TXT)}")
